builtins: make default_to_database_primary_region avoid InternalExecutor

otan opened the pull request:
Resolves #60995 

Using InternalExecutor for default_to_database_primary_region can result
in leasing on the sub-query, which can fail.

Instead, add another interface method onto EvalContext which allows
querying the database descriptor directly. Due to a circular dependency
which is unavoidable without heavy untangling, we are returning an
new interface that wraps RegionConfig. There is precedence for this
kind of thing in around in the `sql` package.

Release justification: low risk change to new functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61045)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
nit on the language of the commit message: it's not that you're avoiding distsql, it's that you're avoiding using the `InternalExecutor` during distributed query execution. 

> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-:-MUHQCnV6PwDTuYqBcoH:b-8cn12u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @otan, and @pbardea)

___
*[pkg/sql/sem/tree/eval.go, line 3016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-MUHQUEK0zKVnn4GiGsg:-MUHQUEK0zKVnn4GiGsh:bhpsi03) ([raw file](https://github.com/cockroachdb/cockroach/blob/286dca279966d862cb7b126a201718c9510e60af/pkg/sql/sem/tree/eval.go#L3016)):*
> ```Go
> 	// Pulling RegionConfig out as a dependency is not an option as the region
> 	// enum ID field would depend on descpb.ID, not alleviating the problem.
> 	CurrentDatabaseRegionConfig(ctx context.Context) (interface{}, error)
> ```

I'd prefer if you made an interface for the `RegionConfig` something like:

```

// DatabaseRegionConfig avoids the circular dependency with descpb.
type DatabaseRegionConfig interface {
    // Provides the string value corresponding to the current region or the
    // default region for the database if this region is not part of the database.
    GetRegionValue(curRegion string) string 
}
```


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, and @pbardea)

___
*[pkg/sql/sem/tree/eval.go, line 3016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-MUHQUEK0zKVnn4GiGsg:-MUHXNkl3SDBkwk-b1Wk:b-qymoe6) ([raw file](https://github.com/cockroachdb/cockroach/blob/286dca279966d862cb7b126a201718c9510e60af/pkg/sql/sem/tree/eval.go#L3016)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd prefer if you made an interface for the `RegionConfig` something like:

```

// DatabaseRegionConfig avoids the circular dependency with descpb.
type DatabaseRegionConfig interface {
    // Provides the string value corresponding to the current region or the
    // default region for the database if this region is not part of the database.
    GetRegionValue(curRegion string) string 
}
```
</blockquote></details>

pretty gross, but alright!


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- I'm content with it like this.

I generally am sad that the builtins can execute statements at all. It's a lot of hoops to jump through going through the executor so this is a big improvement in my book. The number of allocation to execute that builtin was probably very large before and this is going to happen for ~every insert into a small table.

It's about now in the cycle that we might want to run the benchmark on how many allocations we do on a simple kv insert workload -- might be good to extend that to multiregion. The last number I recalled it was in the mid 200s. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-:-MUHqPBC7HFe-CVDQwfk:bmk5rqn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, @otan, and @pbardea)

___
*[pkg/sql/region_util.go, line 778 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-MUHqgk_5BvZGKoep1Cq:-MUHqgk_5BvZGKoep1Cr:bu81u95) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb4e1212ebbc02de06b1f7c4efe35e56e8df17/pkg/sql/region_util.go#L778)):*
> ```Go
> // databaseRegionConfigImpl implements tree.DatabaseRegionConfig.
> type databaseRegionConfigImpl struct {
> 	cfg *descpb.DatabaseDescriptor_RegionConfig
> ```

you'd be able to avoid an allocation on this thing for every call if you just put the methods right on that struct. 
___
*[pkg/sql/sem/tree/eval.go, line 3016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-MUHQUEK0zKVnn4GiGsg:-MUHqEZB38KpBy4XMCGt:b-x27fna) ([raw file](https://github.com/cockroachdb/cockroach/blob/286dca279966d862cb7b126a201718c9510e60af/pkg/sql/sem/tree/eval.go#L3016)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

pretty gross, but alright!
</blockquote></details>

I was thinking you'd implement the interface on the descpb struct but this works I suppose. I find this interface less gross than an empty interface though I'm not enamored by your implementation wrapper.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/tree/eval.go, line 3016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61045#-MUHQUEK0zKVnn4GiGsg:-MUI0VMa125hQ0pErDih:b-bbi82d) ([raw file](https://github.com/cockroachdb/cockroach/blob/286dca279966d862cb7b126a201718c9510e60af/pkg/sql/sem/tree/eval.go#L3016)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I was thinking you'd implement the interface on the descpb struct but this works I suppose. I find this interface less gross than an empty interface though I'm not enamored by your implementation wrapper.
</blockquote></details>

Yeah I didn't like putting it on the regionconfig because it conflicts with existing names and returns different types (i.e. string instead of descpb.RegionName). I'll give a shot putting it directly and see if i like it more.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: make default_to_database_primary_region avoid InternalExecutor

Using InternalExecutor for default_to_database_primary_region can result
in leasing on the sub-query, which can fail.

Instead, add another interface method onto EvalContext which allows
querying the database descriptor directly. Due to a circular dependency
which is unavoidable without heavy untangling, we are returning an
new interface that wraps RegionConfig. There is precedence for this
kind of thing in around in the `sql` package.

Release justification: low risk change to new functionality

Release note: None

ajwerner reviewed (approved):
Yeah I prefer this. The other thing you could have done would have been to create a type with the same structure and cast but that doesn't seem better -- it'll have the same field name conflict problems.

otan commented:
bors r=ajwerner
